src/Spec.ts

```diff
@@ -539,6 +548,240 @@ export default class Spec {
+          } else {
+            this.warn({
+              type: 'node',
+              node: pointer,
+              ruleId,
+              message,
+            });
+          }
+        } else {
+          // @ts-ignore
+          const tree = alg.ecmarkdownTree as ReturnType<typeof parseAlgorithm>;
+          let stepPointer: OrderedListItemNode | UnorderedListItemNode = {
+            sublist: tree.contents,
+          } as OrderedListItemNode;
+          for (const step of path) {
+            stepPointer = stepPointer.sublist!.contents[step];
+          }
+          this.warn({
+            type: 'contents',
+            node: alg,
+            ruleId: 'invocation-return-type',
+            message,
+            nodeRelativeLine: stepPointer.location.start.line,
+            nodeRelativeColumn: stepPointer.location.start.column,
+          });
+        }
+      };
+
+      const consumedAsCompletion = isConsumedAsCompletion(xref);
+      // checks elsewhere ensure that well-formed documents never have a union of completion and non-completion, so checking the first child suffices
+      // TODO: this is for 'a break completion or a throw completion', which is kind of a silly union; maybe address that in some other way?
+      const isCompletion =
+        returnType.kind === 'completion' ||
+        (returnType.kind === 'union' && returnType.types[0].kind === 'completion');
+      if (['Completion', 'ThrowCompletion', 'NormalCompletion'].includes(xref.aoid)) {
+        if (consumedAsCompletion) {
+          // TODO mf to improve message
+          warn(
+            `${xref.aoid} is explicitly a Completion Record and should not be unwrapped or marked`
+          );
+        }
+      } else if (isCompletion && !consumedAsCompletion) {
+        warn(`${xref.aoid} returns a Completion Record, but is not consumed as one`);
+      } else if (!isCompletion && consumedAsCompletion) {
+        warn(`${xref.aoid} does not return a Completion Record, but is consumed as one`);
+      }
+      if (returnType.kind === 'unused' && !isCalledAsPerform(xref, false)) {
+        warn(
+          `${xref.aoid} does not return a value and should only be invoked as \`Perform ${xref.aoid}(...).\``
+        );
+      }
+
+      if (onlyPerformed.has(xref.aoid) && onlyPerformed.get(xref.aoid) !== 'top') {
+        const old = onlyPerformed.get(xref.aoid);
+        const performed = isCalledAsPerform(xref, true);
+        if (!performed) {
+          onlyPerformed.set(xref.aoid, 'top');
+        } else if (old === null) {
+          onlyPerformed.set(xref.aoid, 'only performed');
+        }
+      }
+      if (
+        alwaysAssertedToBeNormal.has(xref.aoid) &&
+        alwaysAssertedToBeNormal.get(xref.aoid) !== 'top'
+      ) {
+        const old = alwaysAssertedToBeNormal.get(xref.aoid);
+        const asserted = isAssertedToBeNormal(xref);
+        if (!asserted) {
+          alwaysAssertedToBeNormal.set(xref.aoid, 'top');
+        } else if (old === null) {
```

Comment:
These `old == null` checks aren't necessary for correctness, but they're fine I guess.

Reply:
I much prefer the `=== null`, since it's the bottom type. Plus it's just confusing to set the type to itself, which omitting the condition or having `old !== 'top'` would do.